### PR TITLE
Make the toplevel build pull-based

### PR DESCRIPTION
**Pull request by jeremiedimino**
This PR changes things so that `Build_system.init` is now only used once at the beginning of the process to tie the knot between `Dune_engine` and `Dune_rules`, and is no longer called on every build in polling mode.

The various values required by `Gen_rules.gen_rules` are now pulled entirely via memo.

We still have a `Import.Main.setup` in `bin/import.ml`, however it is now much simpler and it's only job is to setup the status line.

### CODE REVIEW

File: src/dune_engine/build_system.mli [hunk 1]
```diff
@@ -189,10 +187,19 @@ module For_command_line : sig
...
-
 val cache_teardown : unit -> unit
 
 (** {2 Running a build} *)
 
 val run : (unit -> 'a Memo.Build.t) -> 'a Fiber.t
+
+(** {2 Misc} *)
+
+module Progression : sig
```
Comment (rgrinberg) [inline src/dune_engine/build_system.mli:198]: small nit: `Progress` sounds more natural to me.

File: test/blackbox-tests/test-cases/cxx-flags.t/run.t [hunk 2]
```diff
@@ -47,9 +47,9 @@ With use_standard_c_and_cxx_flags = true
   > EOF
 
 > Check that compiler detection is done
-  $ dune build .dune/ccomp
+  $ dune build .dune/ccomp/ccomp
 
-  $ cat _build/default/.dune/ccomp |
+  $ cat _build/default/.dune/ccomp/ccomp |
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/cxx-flags.t/run.t:52]: Is this change intentional?

File: src/dune_engine/build_system.mli (hunk 1 above)
Comment (jeremiedimino) [reply src/dune_engine/build_system.mli:198]: Sounds good!

File: test/blackbox-tests/test-cases/cxx-flags.t/run.t (hunk 2 above)
Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/cxx-flags.t/run.t:52]: Yeah, I changed the code producing the configurator files to use normal build rules. To force them to be generated eagerly, `gen_rules` requests them in every directory. That created a cycle for these rules, so I moved them in a sub-directory.

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/cxx-flags.t/run.t:52]: BTW, we can now use `Action_builder.action_stdout` for such cases

**Review by rgrinberg — APPROVED** [review]
